API Changes

dralley opened the pull request:
These are the API changes from the single-table-content work that were discussed.

dralley commented on pulpcore/pulpcore/app/models/repository.py:
```diff
@@ -247,6 +247,20 @@ def content(self):
         return Content.objects.filter(version_memberships__in=relationships)
 
+    @property
+    def content_by_type(self):
```
I'm not actually sure this function is necessary or even particularly useful, I may remove it before merging.  It seemed like an obvious thing to have at first but I realized that I can't think of anywhere it would actually be used.

dralley replied:
Should the ```added()``` and ```removed()``` methods be made properties, to match ```content```?  Or should we make ```content``` a method?  The latter would provide maximal flexibility in terms of being able to add functionality later on in the form of kwargs.

dralley commented on pulpcore/pulpcore/app/viewsets/custom_filters.py:
```diff
@@ -60,3 +63,99 @@ class IsoDateTimeFilter(DateTimeFilter):
         super().__init__(*args, **kwargs)
+
+
+class ContentRepositoryVersionFilter(Filter):
```
I want to DRY this but when I tried, it wasn't passing the "value" correctly.  I'll get to it later if I can.

dralley commented on pulpcore/pulpcore/app/models/repository.py:
```diff
@@ -247,6 +249,62 @@ def content(self):
         )
         return Content.objects.filter(version_memberships__in=relationships)
 
+    @property
```
These 3 properties are probably temporary, pending a better solution.  This is ugly but it was fairly simple for an initial implementation.

dralley commented on pulpcore/pulpcore/app/viewsets/__init__.py:
```diff
@@ -1,10 +1,16 @@
 from .base import AsyncRemoveMixin, AsyncUpdateMixin, BaseFilterSet, NamedModelViewSet  # noqa
+# These need to be imported prior to ContentViewSet - do not shuffle this by alphabetical order
+from .custom_filters import (  # noqa
```
I struggled a little bit with circular imports here -- should we just move these to viewsets.content since they're only used in that one place?

dralley commented on pulpcore/pulpcore/app/viewsets/custom_filters.py:
```diff
@@ -60,3 +63,101 @@ class IsoDateTimeFilter(DateTimeFilter):
         super().__init__(*args, **kwargs)
+
+
+class RepoVersionHrefFilter(Filter):
```
Should I rename this to "BaseRepoVersionHrefFilter" since it's meant to be subclassed and not used directly?

dralley commented on pulpcore/tests/functional/api/using_plugin/test_repo_versions.py:
```diff
@@ -110,13 +110,13 @@ def test_02_sync_content(self):
         self.assertIsNotNone(repo['_latest_version_href'])
 
-        content = get_content(repo)
+        content = get_file_content(repo)
```
As mentioned earlier, we could just always do ```get_content(repo)['file']``` instead.  Open to changing this if QE would prefer it that way.

dralley commented:
TODO: Docs changes.  I'll handle that after the code review is hashed out.

daviddavis commented on plugin/pulpcore/plugin/viewsets.py:
```diff
@@ -2,7 +2,7 @@
 
 # Import Viewsets in platform that are potentially useful to plugin writers
-from pulpcore.app.viewsets import (  # NOQA
+from pulpcore.app.viewsets import (  # noqa
```
I fixed this [here](https://github.com/pulp/pulpcore-plugin/pull/3).

dkliban commented:
Perhaps we should just get rid of 'content', 'content_added', and 'content_removed' from the RepsositoryVersion serializer. If users want this in the future we can always add it.

dralley commented on pulpcore/app/serializers/repository.py:
```diff
@@ -407,25 +407,21 @@ class RepositoryVersionSerializer(ModelSerializer, NestedHyperlinkedModelSeriali
-    _content_href = NestedIdentityField(
-        view_name='versions-content',
-        lookup_field='number', parent_lookup_kwargs={'repository_pk': 'repository__pk'},
+    content = serializers.SerializerMethodField(
```
As I mentioned on IRC, we should be consider leaving some space for future expansion should it be necessary. 

I suppose if we ever needed to add back in the content urls, we could use "_hrefs" as we did previously.  But we could consider other alternatives.

bmbouter commented:
re: if urls should be included in responses or not. I think they should be because that is more usable for humans even though the bindings can't consume it in any reasonable way.

dralley commented:
@dkliban @bmbouter Alright, I've added back URLs and also added counts for added and removed content.  So we can subtract from this if we decide we don't want something.

dralley commented on pulpcore/app/serializers/repository.py:
```diff
@@ -435,12 +421,154 @@ class RepositoryVersionSerializer(ModelSerializer, NestedHyperlinkedModelSeriali
+        Returns:
+            dict: of {<type>: <count>}
+        """
+        annotated = obj.content.values('type').annotate(count=Count('type'))
```
We should consider whether this should be made a method like ```added()``` and ```removed()``` to be more consistent, and also leave open the possibility of adding kwargs later.  But that can be a different PR.

dralley commented on pulpcore/app/serializers/repository.py:
```diff
@@ -435,12 +421,154 @@ class RepositoryVersionSerializer(ModelSerializer, NestedHyperlinkedModelSeriali
+        content_urls = {}
+
+        for ctype in obj.content.values_list('type', flat=True):
+            ctype_model = obj.content.filter(type=ctype).first().cast().__class__
```
It would be good if we could find a better way to do this
